Add back JUnit style reports via the mocha-multi-reporters plugin

d3r3kk opened the pull request:
For #4759

- [x] Add back the JUnit output for mocha tests run programmatically as well.
  - [x] Add to testRunner.ts when configuring the Mocha object.
  - [x] Incorporate environment variable flag MOCHA_JUNIT_REPORTER to signal when to turn it on.
- [x] Upload multiple test run JUnit xml files if multiple are produced.
- [x] Keep coloured console output intact and unbuffered as it is without any reporters.

---


- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] ~Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!)~
- [x] Has sufficient logging.
- [x] Has telemetry for enhancements.
- [x] Unit tests & system/integration tests are added/updated
- [x] ~[Test plan](https://github.com/Microsoft/vscode-python/blob/master/.github/test_plan.md) is updated as appropriate~
- [x] [`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed)
- [x] ~The wiki is updated with any design decisions/details.~

DonJayamanne commented on src/test/testRunner.ts:
```diff
@@ -116,7 +117,7 @@ export function run(testsRoot: string, callback: TestCallback): void {
             }
             try {
-                files.forEach(file => mocha.addFile(path.join(testsRoot, file)));
+                files.splice(0, 10).forEach(file => mocha.addFile(path.join(testsRoot, file)));
```
Why are we removing the first 10 files?

d3r3kk replied:
WIP

DonJayamanne commented on build/ci/vscode-python-pr-validation.yaml:
```diff
@@ -18,6 +18,7 @@ variables:
   PythonVersion: '3.7'
   NodeVersion: '8.11.2'
   NpmVersion: 'latest'
+  MOCHA_FILE: ${{ format('{0}/test-junit.xml', variables['Build.ArtifactStagingDirectory']) }}
```
Where is this variable used? Couldn't find a reference to it, could be mistaken.

d3r3kk replied:
MOCHA_FILE is used by the `mocha-junit-reporter` to tell it where to place the output file.

Setting it here, in the build definition ensures it's always in a known place.

`Build.ArtifactStagingDirectory` is as apt a place as any to put it.

But this is likely not final, WIP.

DonJayamanne commented:
Oops, sorry, didn't realize this was WIP

d3r3kk commented:
**Note**: Once this gets merged in it will fix some of the problems we're seeing in our CI builds as well (internal tool tests failing on non-3.7 Python).

ericsnowcurrently commented on src/test/index.ts:
```diff
@@ -40,14 +38,13 @@ const options: testRunner.SetupOptions & { retries: number } = {
 if (MOCHA_REPORTER_JUNIT) {
-    options.reporter = MOCHA_CI_REPORTER_ID;
+    options.reporter = 'mocha-multi-reporters';
     options.reporterOptions = {
```
Why can't we use `build/.mocha-multi-reporters.config`?

d3r3kk replied:
Because AFAIK that file doesn't get carried over into the `out/test` folders once we build this thing... that's where we are running from in this scenario, in a binaries-only step. I could be wrong, but I am building on the call to `configure` here, found in `testRunner.ts` which configures `mocha` this way.

ericsnowcurrently commented on src/test/index.ts:
```diff
@@ -40,14 +38,13 @@ const options: testRunner.SetupOptions & { retries: number } = {
-// 'MOCHA_REPORTER_JUNIT' is defined, further control is afforded
-// by other 'MOCHA_CI_...' variables. See constants.ts for info.
+// If the `MOCHA_REPORTER_JUNIT` env var is true, set up the CI reporter for
+// reporting to the console as well as to the JUnit XML file (defaults to
```
This makes it sound like it doesn't normally report to the console. :)

d3r3kk replied:
Gotcha. English is hard.

ericsnowcurrently commented on src/test/common/moduleInstaller.test.ts:
```diff
@@ -87,7 +91,7 @@ suite('Module Installer', () => {
             // If resource is provided, then ensure we do not invoke without the resource.
             mockTerminalFactory.setup(t => t.getTerminalService(TypeMoq.It.isAny()))
-                .callback(passedInResource => expect(passedInResource).to.be.equal(resource))
+                .callback(passedInResource => chai.expect(passedInResource).to.be.equal(resource))
```
Did you mean to change this?  I don't see where "chai" is imported.

ericsnowcurrently replied:
Obviously it's fine, or CI would fail. :)

d3r3kk replied:
Nice catch! And man, I can't sort this issue with `chai` out that well. I was only running 10 of the files in my iterations on this change, and that caused this file to fail with `unknown property 'eventually'...` errors due to the `chai/chai-as-promised` setup being incomplete. 

The only thing I could come up with is that mocha must 'coalesce' files it is going to run, and in doing so pulls in the necessary imports from elsewhere?

But anyways, this oversight didn't cause any failures in my PR validation...

ericsnowcurrently commented on build/ci/vscode-python-pr-validation.yaml:
```diff
@@ -40,7 +42,7 @@ jobs:
       'Win-Py3.7 Unit':
         VMImageName: 'vs2017-win2016'
-        TestsToRun: 'runHygiene, testUnitTests, pythonUnitTests'
+        TestsToRun: 'runHygiene, testUnitTests, pythonUnitTests, pythonInternalTools'
```
Only for Windows?

d3r3kk replied:
It could be on Linux or Mac too - if it's important I'll leave it up to you fine folks to put it wherever you like (this PR is just making it possible to put anywhere, I chose Windows as that's likely where it will be run from, but Mac was a close runner up).

d3r3kk replied:
Ah I'll just add it...

d3r3kk commented:
@karrtikr I split out the smoke tests in this PR after Eric made some suggestions.

@DonJayamanne I split out testSingleWorkspace as you suggested, let's see if it speeds things up or not.

d3r3kk commented:
PR #4869 handles the unit-test failures seen in the PR Validation for this change.
